### PR TITLE
fix: cli is not a library!

### PR DESCRIPTION
**Pull request by rb-determined-ai**
Direct calls into the CLI are problematic, because the CLI uses a variety of authentication techniques, some of which are only configured by the cli's __main__() codepath.  That means that cli auth and cert singletons have been configured all over e2e tests in order to support direct CLI calls.

Avoiding direct calls to cli functions is a step in the direction of cleaning up our authentication story.

### CODE REVIEW

File: harness/determined/common/api/_util.py
Comment (rb-determined-ai) [inline harness/determined/common/api/_util.py:151]: wanna call this out:

I know we're using api.errors.BadRequestException to surface a nice user-facing error message.

I'm not happy with the way we're using it, but wanted to know if you had a particular idea as to how to raise a not-overly-CLI-specific error in a "helper function" that, let's be real, is mostly only for the CLI.  And a couple e2e tests.

File: e2e_tests/tests/cluster/test_logging.py
Comment (hamidzr) [inline e2e_tests/tests/cluster/test_logging.py:66]: hmm why?

**Review by hamidzr — COMMENTED** [review]
looks good I don't have much to add here I'll let mlsys review unless there is a specific part you want me to look at

File: e2e_tests/tests/cluster/test_logging.py
Comment (rb-determined-ai) [reply e2e_tests/tests/cluster/test_logging.py:66]: because the CLI is not a library.  It should _never_ be imported from outside the `cli/` directory.

Importing it for a few static strings is a violation of the application-nature of the CLI, with no particular benefit.  I don't think these strings are going to change in the next 10 years.

Comment (hamidzr) [reply e2e_tests/tests/cluster/test_logging.py:66]: I didn't check where the `command` namespace was coming from. makes sense.

File: e2e_tests/tests/cluster/test_rbac.py
Comment (wes-turner) [inline e2e_tests/tests/cluster/test_rbac.py:65]: Excision of `determined.cli.rbac.assign_role` into `determined.common.api.assign_role` looks great, as does its use here. `api.assign_role` seems like a pretty reasonable place to call something from that fronts the backend. But there's something about it living in `determined.common.api._util` that feels wrong to me. It's a nit either way, and who really cares what _file_ the function is in, but if you think that it'll be helpful to us to someday have structure to that directory, I think now is a good time to start it.

File: harness/determined/common/api/_util.py
Comment (wes-turner) [inline harness/determined/common/api/_util.py:255]: I don't get what's going on with this function. Could it be part of your cleanup?

**Review by wes-turner — APPROVED** [review]
LGTM. Thanks for making this better.

File: harness/determined/common/api/_util.py
Comment (rb-determined-ai) [reply harness/determined/common/api/_util.py:161]: Ok, it sounds like since neither you nor I are sure what to do here, I'll just let it be.

I can explain why I don't like it in its current state, but I can't explain what would be better, and since at the end of the day the current form makes for a very nice user experience, changing it now feels like churn that would be unlikely to "stick" long-term.  Better to wait until we have a clearer motivating example as to what should be done, and fix according to that motivating example.

File: e2e_tests/tests/cluster/test_rbac.py
Comment (rb-determined-ai) [reply e2e_tests/tests/cluster/test_rbac.py:65]: I really don't care what file the function is in.  If you want me to put it in a different file, lmk what to name it and I'll do it.

I do like having it in the `api` namespace, and this is a case where I really like the separation between how code is called and where code is defined.

File: harness/determined/common/api/_util.py
Comment (rb-determined-ai) [reply harness/determined/common/api/_util.py:255]: Yes, I'll look into it.  I didn't ask that question earlier, I just shuffled it around.

File: harness/determined/common/api/_util.py
Comment (wes-turner) [reply harness/determined/common/api/_util.py:161]: Word. I support that approach.

File: e2e_tests/tests/cluster/test_rbac.py
Comment (wes-turner) [reply e2e_tests/tests/cluster/test_rbac.py:65]: > I do like having it in the api namespace, and this is a case where I really like the separation between how code is called and where code is defined.

I've not had cause to differentiate the two in the past, but I agree. It's really nice, here.

This comment applies to many of the other moved functions, too. How about `determined/common/api/_rbac.py`?

File: harness/determined/common/api/_util.py
Comment (rb-determined-ai) [reply harness/determined/common/api/_util.py:255]: looked into it.  It just helps users remember that 404 errors in rbac cases might be permissions errors.

I promoted the comment to a docstring and added more explanation.

File: e2e_tests/tests/cluster/test_rbac.py
Comment (rb-determined-ai) [reply e2e_tests/tests/cluster/test_rbac.py:65]: I think that sounds like a great name.

I put all the functions I added there, since they're all rbac-related.  Even not_found_errs is rbac-related.

File: harness/determined/common/api/_util.py
Comment (wes-turner) [reply harness/determined/common/api/_util.py]: The docstring is a better place to put the help text, for sure. And thanks for figuring out what the function is for.

I still have to do more work to understand what the function does than I want a person to have to do. Something seems off about the design. Bleh.

But anyway, I think bigger changes are outside the scope of this PR. So maybe here just a suggestion like:
```
"""Construct a NotFoundException from passed strings.

This function creates NotFoundExceptions with the same syntax as the NotFoundErrs from api/errors.go so that messages from both layers look the same. If RBAC is enabled, the exceptions underlying the 404s that this function constructs might be the result of a permissions error. In that event, this function appends a suggestion to the constructed exception string that the user to check permissions.
```

The type mismatch between the actual return type and its superclass is a little weird. That'd be an easy fix if you want to make it.

File: harness/determined/common/api/_rbac.py
Comment (wes-turner) [inline harness/determined/common/api/_rbac.py:25]: Don't forget to either remove me or turn this into a TODO and/or ticket.

File: e2e_tests/tests/cluster/test_rbac.py
Comment (wes-turner) [reply e2e_tests/tests/cluster/test_rbac.py:65]: 👍  I like it.

File: harness/determined/common/api/_rbac.py
Comment (rb-determined-ai) [reply harness/determined/common/api/_rbac.py:25]: ah yes, thank you

**Review by stoksc — APPROVED** [review]
backend lgtm as long as tests pass

**Comment by rb-determined-ai** [comment]
status update: this is blocked on bringing the EE changes in e2e_tests into the OSS repo.

We can't solve the rebase problem for the whole world, but we can solve it for this directory!

**Comment by rb-determined-ai** [comment]
status update: EE changes to e2e_tests are finally in OSS now, and this PR can proceed as planned.

File: harness/determined/common/api/_rbac.py
Comment (wes-turner) [inline harness/determined/common/api/_rbac.py:17]: I don't like this method. It's got all kinds of weird. It's not your job, but do you want to be the person on the other side of the git blame?

I think it should be split into:
```python
create_user_assignment_request(session, role, workspace, username) -> bindings.v1UserRoleAssignment
create_group_assignment_request(session, role, workspace, group) -> bindings.v1GroupRoleAssignment
```

**Review by wes-turner — APPROVED** [review]
New changes yes.

File: harness/determined/common/api/_rbac.py
Comment (rb-determined-ai) [reply harness/determined/common/api/_rbac.py:17]: lol, you are exactly right.  I didn't even read the function body.

**Comment by rb-determined-ai** [comment]
I removed a review request from the web team, seems to have been from an old force-push-to-master incident.

**Comment by rb-determined-ai** [comment]
Everything passed except some known flakes.  Same story [over in the EE repo](https://github.com/determined-ai/determined-ee/pull/1119).  Force-merging.

**Pull request merged**
